File: bindings/haskell/src/ProvenServers/Graphdb.hs
-- SPDX-License-Identifier: MPL-2.0
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- | Graph Database types for the proven-servers ABI.
--
-- All tag values match the Idris2 ABI discriminants exactly.

module ProvenServers.Graphdb
  (
    graphdbPort
  , ElementType(..)
  , elementTypeToTag
  , elementTypeFromTag
  , QueryLanguage(..)
  , queryLanguageToTag
  , queryLanguageFromTag
  , TraversalStrategy(..)
  , traversalStrategyToTag
  , traversalStrategyFromTag
  , Consistency(..)
  , consistencyToTag
  , consistencyFromTag
  , ErrorCode(..)
  , errorCodeToTag
  , errorCodeFromTag
  , SessionState(..)
  , sessionStateToTag
  , sessionStateFromTag
  ) where

import Data.Word (Word16, Word8)

-- | Standard Bolt protocol port.
graphdbPort :: Word16
graphdbPort = 7687

-- ---------------------------------------------------------------------------
-- ElementType
-- ---------------------------------------------------------------------------

-- | Standard Bolt protocol port.
--
-- Tags 0-4 (5 constructors).
data ElementType
  = Node  -- ^ Node (tag 0).
  | Edge  -- ^ Edge (tag 1).
  | Property  -- ^ Property (tag 2).
  | Label  -- ^ Label (tag 3).
  | Index  -- ^ Index (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ElementType' to its ABI tag value.
elementTypeToTag :: ElementType -> Word8
elementTypeToTag = fromIntegral . fromEnum

-- | Decode a 'ElementType' from its ABI tag value.
elementTypeFromTag :: Word8 -> Maybe ElementType
elementTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ElementType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- QueryLanguage
-- ---------------------------------------------------------------------------

-- | Graph query languages.
--
-- Tags 0-3 (4 constructors).
data QueryLanguage
  = Cypher  -- ^ Cypher (tag 0).
  | Gremlin  -- ^ Gremlin (tag 1).
  | Sparql  -- ^ SPARQL (tag 2).
  | GraphQl  -- ^ GraphQL (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'QueryLanguage' to its ABI tag value.
queryLanguageToTag :: QueryLanguage -> Word8
queryLanguageToTag = fromIntegral . fromEnum

-- | Decode a 'QueryLanguage' from its ABI tag value.
queryLanguageFromTag :: Word8 -> Maybe QueryLanguage
queryLanguageFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: QueryLanguage)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- TraversalStrategy
-- ---------------------------------------------------------------------------

-- | Graph traversal strategies.
--
-- Tags 0-4 (5 constructors).
data TraversalStrategy
  = Bfs  -- ^ Breadth-first search (tag 0).
  | Dfs  -- ^ Depth-first search (tag 1).
  | Dijkstra  -- ^ Dijkstra (tag 2).
  | AStar  -- ^ A* (tag 3).
  | Random  -- ^ Random (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'TraversalStrategy' to its ABI tag value.
traversalStrategyToTag :: TraversalStrategy -> Word8
traversalStrategyToTag = fromIntegral . fromEnum

-- | Decode a 'TraversalStrategy' from its ABI tag value.
traversalStrategyFromTag :: Word8 -> Maybe TraversalStrategy
traversalStrategyFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: TraversalStrategy)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- Consistency
-- ---------------------------------------------------------------------------

-- | Consistency levels.
--
-- Tags 0-3 (4 constructors).
data Consistency
  = Strong  -- ^ Strong (tag 0).
  | Eventual  -- ^ Eventual (tag 1).
  | Session  -- ^ Session (tag 2).
  | Causal  -- ^ Causal (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'Consistency' to its ABI tag value.
consistencyToTag :: Consistency -> Word8
consistencyToTag = fromIntegral . fromEnum

-- | Decode a 'Consistency' from its ABI tag value.
consistencyFromTag :: Word8 -> Maybe Consistency
consistencyFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: Consistency)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- ErrorCode
-- ---------------------------------------------------------------------------

-- | Graph database error codes.
--
-- Tags 0-6 (7 constructors).
data ErrorCode
  = SyntaxError  -- ^ SyntaxError (tag 0).
  | NodeNotFound  -- ^ NodeNotFound (tag 1).
  | EdgeNotFound  -- ^ EdgeNotFound (tag 2).
  | ConstraintViolation  -- ^ ConstraintViolation (tag 3).
  | IndexExists  -- ^ IndexExists (tag 4).
  | TransactionConflict  -- ^ TransactionConflict (tag 5).
  | OutOfMemory  -- ^ OutOfMemory (tag 6).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'ErrorCode' to its ABI tag value.
errorCodeToTag :: ErrorCode -> Word8
errorCodeToTag = fromIntegral . fromEnum

-- | Decode a 'ErrorCode' from its ABI tag value.
errorCodeFromTag :: Word8 -> Maybe ErrorCode
errorCodeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: ErrorCode)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- SessionState
-- ---------------------------------------------------------------------------

-- | Graph database session states.
--
-- Tags 0-4 (5 constructors).
data SessionState
  = Idle  -- ^ Idle (tag 0).
  | Connected  -- ^ Connected (tag 1).
  | Querying  -- ^ Querying (tag 2).
  | Traversing  -- ^ Traversing (tag 3).
  | Disconnecting  -- ^ Disconnecting (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'SessionState' to its ABI tag value.
sessionStateToTag :: SessionState -> Word8
sessionStateToTag = fromIntegral . fromEnum

-- | Decode a 'SessionState' from its ABI tag value.
sessionStateFromTag :: Word8 -> Maybe SessionState
sessionStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: SessionState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

File: bindings/elixir/lib/proven_servers/graphdb.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Graphdb do
  @moduledoc """
  Graph Database types for the proven-servers ABI.
  
  Formally verified graph database types.
  Mirrors the Idris2 module `GraphdbABI.Types`.
  
  - `ElementType` -- Graph element types.
  - `QueryLanguage` -- Graph query languages.
  - `TraversalStrategy` -- Graph traversal strategies.
  - `Consistency` -- Consistency levels.
  - `ErrorCode` -- Graph database error codes.
  - `SessionState` -- Graph database session states.
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard Bolt protocol port."
  @spec graphdb_port() :: non_neg_integer()
  def graphdb_port, do: 7687

  # ===========================================================================
  # ElementType (tags 0-4)
  # ===========================================================================

  @typedoc """
  ElementType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type element_type :: :node | :edge | :property | :label | :index

  @element_type_tags %{
    node: 0,
    edge: 1,
    property: 2,
    label: 3,
    index: 4,
  }

  @tag_to_element_type Map.new(@element_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ElementType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.element_type_from_tag(0)
      {:ok, :node}
  """
  @spec element_type_from_tag(non_neg_integer()) :: {:ok, element_type()} | :error
  def element_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_element_type, tag)}
  end

  def element_type_from_tag(_tag), do: :error

  @doc """
  Encode a `ElementType` to the C-ABI tag value.
  """
  @spec element_type_to_tag(element_type()) :: non_neg_integer()
  def element_type_to_tag(val) when is_map_key(@element_type_tags, val) do
    Map.fetch!(@element_type_tags, val)
  end

  @doc """
  All `ElementType` variants in tag order.
  """
  @spec all_element_types() :: [element_type()]
  def all_element_types, do: [:node, :edge, :property, :label, :index]

  # ===========================================================================
  # QueryLanguage (tags 0-3)
  # ===========================================================================

  @typedoc """
  QueryLanguage types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type query_language :: :cypher | :gremlin | :sparql | :graph_ql

  @query_language_tags %{
    cypher: 0,
    gremlin: 1,
    sparql: 2,
    graph_ql: 3,
  }

  @tag_to_query_language Map.new(@query_language_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `QueryLanguage` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.query_language_from_tag(0)
      {:ok, :cypher}
  """
  @spec query_language_from_tag(non_neg_integer()) :: {:ok, query_language()} | :error
  def query_language_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_query_language, tag)}
  end

  def query_language_from_tag(_tag), do: :error

  @doc """
  Encode a `QueryLanguage` to the C-ABI tag value.
  """
  @spec query_language_to_tag(query_language()) :: non_neg_integer()
  def query_language_to_tag(val) when is_map_key(@query_language_tags, val) do
    Map.fetch!(@query_language_tags, val)
  end

  @doc """
  All `QueryLanguage` variants in tag order.
  """
  @spec all_query_languages() :: [query_language()]
  def all_query_languages, do: [:cypher, :gremlin, :sparql, :graph_ql]

  # ===========================================================================
  # TraversalStrategy (tags 0-4)
  # ===========================================================================

  @typedoc """
  TraversalStrategy types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type traversal_strategy :: :bfs | :dfs | :dijkstra | :a_star | :random

  @traversal_strategy_tags %{
    bfs: 0,
    dfs: 1,
    dijkstra: 2,
    a_star: 3,
    random: 4,
  }

  @tag_to_traversal_strategy Map.new(@traversal_strategy_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `TraversalStrategy` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.traversal_strategy_from_tag(0)
      {:ok, :bfs}
  """
  @spec traversal_strategy_from_tag(non_neg_integer()) :: {:ok, traversal_strategy()} | :error
  def traversal_strategy_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_traversal_strategy, tag)}
  end

  def traversal_strategy_from_tag(_tag), do: :error

  @doc """
  Encode a `TraversalStrategy` to the C-ABI tag value.
  """
  @spec traversal_strategy_to_tag(traversal_strategy()) :: non_neg_integer()
  def traversal_strategy_to_tag(val) when is_map_key(@traversal_strategy_tags, val) do
    Map.fetch!(@traversal_strategy_tags, val)
  end

  @doc """
  All `TraversalStrategy` variants in tag order.
  """
  @spec all_traversal_strategys() :: [traversal_strategy()]
  def all_traversal_strategys, do: [:bfs, :dfs, :dijkstra, :a_star, :random]

  # ===========================================================================
  # Consistency (tags 0-3)
  # ===========================================================================

  @typedoc """
  Consistency types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type consistency :: :strong | :eventual | :session | :causal

  @consistency_tags %{
    strong: 0,
    eventual: 1,
    session: 2,
    causal: 3,
  }

  @tag_to_consistency Map.new(@consistency_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Consistency` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.consistency_from_tag(0)
      {:ok, :strong}
  """
  @spec consistency_from_tag(non_neg_integer()) :: {:ok, consistency()} | :error
  def consistency_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_consistency, tag)}
  end

  def consistency_from_tag(_tag), do: :error

  @doc """
  Encode a `Consistency` to the C-ABI tag value.
  """
  @spec consistency_to_tag(consistency()) :: non_neg_integer()
  def consistency_to_tag(val) when is_map_key(@consistency_tags, val) do
    Map.fetch!(@consistency_tags, val)
  end

  @doc """
  All `Consistency` variants in tag order.
  """
  @spec all_consistencys() :: [consistency()]
  def all_consistencys, do: [:strong, :eventual, :session, :causal]

  # ===========================================================================
  # ErrorCode (tags 0-6)
  # ===========================================================================

  @typedoc """
  ErrorCode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type error_code ::
          :syntax_error
          | :node_not_found
          | :edge_not_found
          | :constraint_violation
          | :index_exists
          | :transaction_conflict
          | :out_of_memory

  @error_code_tags %{
    syntax_error: 0,
    node_not_found: 1,
    edge_not_found: 2,
    constraint_violation: 3,
    index_exists: 4,
    transaction_conflict: 5,
    out_of_memory: 6,
  }

  @tag_to_error_code Map.new(@error_code_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ErrorCode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..6, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.error_code_from_tag(0)
      {:ok, :syntax_error}
  """
  @spec error_code_from_tag(non_neg_integer()) :: {:ok, error_code()} | :error
  def error_code_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_error_code, tag)}
  end

  def error_code_from_tag(_tag), do: :error

  @doc """
  Encode a `ErrorCode` to the C-ABI tag value.
  """
  @spec error_code_to_tag(error_code()) :: non_neg_integer()
  def error_code_to_tag(val) when is_map_key(@error_code_tags, val) do
    Map.fetch!(@error_code_tags, val)
  end

  @doc """
  All `ErrorCode` variants in tag order.
  """
  @spec all_error_codes() :: [error_code()]
  def all_error_codes do
    [
      :syntax_error, :node_not_found, :edge_not_found, :constraint_violation,
      :index_exists, :transaction_conflict, :out_of_memory
    ]
  end

  # ===========================================================================
  # SessionState (tags 0-4)
  # ===========================================================================

  @typedoc """
  SessionState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type session_state :: :idle | :connected | :querying | :traversing | :disconnecting

  @session_state_tags %{
    idle: 0,
    connected: 1,
    querying: 2,
    traversing: 3,
    disconnecting: 4,
  }

  @tag_to_session_state Map.new(@session_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SessionState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Graphdb.session_state_from_tag(0)
      {:ok, :idle}
  """
  @spec session_state_from_tag(non_neg_integer()) :: {:ok, session_state()} | :error
  def session_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_session_state, tag)}
  end

  def session_state_from_tag(_tag), do: :error

  @doc """
  Encode a `SessionState` to the C-ABI tag value.
  """
  @spec session_state_to_tag(session_state()) :: non_neg_integer()
  def session_state_to_tag(val) when is_map_key(@session_state_tags, val) do
    Map.fetch!(@session_state_tags, val)
  end

  @doc """
  All `SessionState` variants in tag order.
  """
  @spec all_session_states() :: [session_state()]
  def all_session_states, do: [:idle, :connected, :querying, :traversing, :disconnecting]

end
